File: SCL/Datastruct/TrieTree/最大异或对.py
if __name__ == "__main__":
    n = int(input())
    N = n + 5
    M = 30 * N
    son = [[0,0] for _ in range(M)]
    idx = 0
    
    def insert(x):
        p = 0
        global idx
        for i in range(30,-1,-1):
            u = x >> i & 1
            if (son[p][u] == 0):
                idx += 1
                son[p][u] = idx
            p = son[p][u]
    
    def query(x):
        p = 0
        res = 0
        for i in range(30,-1,-1):
            u = x >> i & 1
            if(son[p][1-u] != 0):
               res = res * 2 + 1-u
               p = son[p][1- u]
            else:
                 res = res * 2 + u
                 p = son[p][u]
        return res 
    
    
    res = 0
    a = list(map(int,input().split()))
    for x in a :
        insert(x)
        t = query(x)
        res = max(res,x ^ t)
        
    print(res)

File: Datastruct/TrieTree/最大异或对.cpp
#include <iostream>
#include <algorithm>

//首先想最暴力的解法，就是两重循环，逐个判断每对异或之后找到最大的
//优化的想法来自于这个数的二进制表示，我们用trie树存下每一个出现的数，然后从高位到低位，尽量选择每一位与他不相同的数
//理解异或最简单的方法就是把他理解成不进位加法
//这样我们就可以拿到每个数的最大异或，之后比较一下就行了
using namespace std;

const int N = 1e5 +10,M = 31 * N;

int son[M][2],idx;

void insert(int x){
    int p = 0;
    for (int i = 30;i >= 0;i--){
        int u = x >> i & 1;
        if (!son[p][u]) son[p][u] = ++idx;
        p = son[p][u];
    }
}

int query(int x){
    int p = 0,res = 0;
    for(int i = 30;i >= 0;i--){
        int u = x >> i & 1;
        if(son[p][!u]){
            res = res * 2 + !u;
            p = son[p][!u];
        }else{
            res = res * 2 + u;
            p = son[p][u];
        }
    }
    return res;
}

int main(){
    int n;
    cin >> n;
    int res = 0;
    for(int i = 0;i < n;i++){
        int cur;
        scanf("%d",&cur);
        insert(cur);
        int t = query(cur);
        res = max(res,cur ^ t);
    }
    cout << res << endl;
    return 0;
}